### PR TITLE
M3 query language grammar

### PR DESCRIPTION
**Pull request by nikunjgit**

### CODE REVIEW

**Comment by andrewmains12** [comment]
Are there tests that should come with this?

**Comment by nikunjgit** [comment]
src/query/parser/m3ql/grammar_test.go ?

File: src/query/parser/m3ql/grammar_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGrammar(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	sBuilder := NewMockscriptBuilder(ctrl)
+	sBuilder.EXPECT().newPipeline()
```
[the remark this answers is not in the document]
Comment (nikunjgit) [reply src/query/parser/m3ql/grammar_test.go]: don't think you can chain like that. Note that it's a mock and we never implement a concrete type

File: src/query/generated/mocks/generate.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// mockgen rules for generating mocks for unexported interfaces (file mode).
+//go:generate sh -c "mockgen -package=m3ql -destination=$GOPATH/src/github.com/m3db/m3/src/query/parser/m3ql/types_mock.go -source=$GOPATH/src/github.com/m3db/m3/src/query/parser/m3ql/types.go"
+
+package mocks
```
Comment (arnikola) [inline src/query/generated/mocks/generate.go]: I added [some generators to mockgen](https://github.com/m3db/m3/blob/0bd9aba9c9616593b79b5945c96657d3c23add58/src/dbnode/generated/mocks/generate.go#L28) a while ago which should probably be moved into here, mind taking care of it?

**Comment by arnikola** [comment]
Make sure you add all the generated files after running `make all-gen-query`, otherwise CI will fail

File: src/query/parser/m3ql/grammar_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewmains12) [reply src/query/parser/m3ql/grammar_test.go]: Ah yeah, I was assuming there existed a concrete implementation for it somewhere. This'll work then :)

File: src/query/generated/mocks/generate.go (hunk 2 above)
Comment (nikunjgit) [reply src/query/generated/mocks/generate.go]: okay okay

**Pull request merged**